Issue with Java HttpUrlConnection's Accept header

From Lucas:

$ nc hub-v2.svc.staging 80
GET /channel/kratzen_ams/2015/03/19/03/47/55/655/ujs8Nb HTTP/1.1
User-Agent: Java/1.7.0_71
Host: hub-v2.svc.staging
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2

Stack trace from logs:
java.lang.IllegalArgumentException: java.text.ParseException: Expected separator '/' instead of ';'
	at com.sun.jersey.core.header.MediaTypes.createMediaTypes(MediaTypes.java:234) ~[jersey-bundle-1.19.jar:1.19]
	at com.flightstats.hub.channel.ChannelContentResource.contentTypeIsNotCompatible(ChannelContentResource.java:324) ~[hub-2015-03-18.7.jar:na]
	at com.flightstats.hub.channel.ChannelContentResource.getValue(ChannelContentResource.java:186) ~[hub-2015-03-18.7.jar:na]
--
Caused by: java.text.ParseException: Expected separator '/' instead of ';'
	at com.sun.jersey.core.header.reader.HttpHeaderReader.nextSeparator(HttpHeaderReader.java:121) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.impl.provider.header.MediaTypeProvider.valueOf(MediaTypeProvider.java:90) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.header.reader.HttpHeaderReader$2.create(HttpHeaderReader.java:320) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.header.reader.HttpHeaderReader$2.create(HttpHeaderReader.java:318) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.header.reader.HttpHeaderReader.readList(HttpHeaderReader.java:481) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.header.reader.HttpHeaderReader.readMediaTypes(HttpHeaderReader.java:325) ~[jersey-bundle-1.19.jar:1.19]
	at com.sun.jersey.core.header.MediaTypes.createMediaTypes(MediaTypes.java:228) ~[jersey-bundle-1.19.jar:1.19]

--- a/src/main/java/com/flightstats/hub/channel/ChannelContentResource.java
+++ b/src/main/java/com/flightstats/hub/channel/ChannelContentResource.java
@@ -16,6 +16,7 @@ import com.google.common.io.ByteStreams;
 import com.google.inject.Inject;
 import com.sun.jersey.api.Responses;
 import com.sun.jersey.core.header.MediaTypes;
+import net.logstash.logback.encoder.org.apache.commons.lang.StringUtils;
 import org.joda.time.DateTime;
 import org.joda.time.DateTimeZone;
 import org.joda.time.format.DateTimeFormatter;
@@ -29,7 +30,9 @@ import javax.ws.rs.core.Response;
 import javax.ws.rs.core.StreamingOutput;
 import javax.ws.rs.core.UriInfo;
 import java.net.URI;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 
 import static com.flightstats.hub.util.TimeUtil.*;
@@ -319,10 +322,17 @@ public class ChannelContentResource {
         return MediaType.APPLICATION_OCTET_STREAM_TYPE;
     }
 
-    private boolean contentTypeIsNotCompatible(String acceptHeader, final MediaType actualContentType) {
-        List<MediaType> acceptableContentTypes = acceptHeader != null ?
-                MediaTypes.createMediaTypes(acceptHeader.split(",")) :
-                MediaTypes.GENERAL_MEDIA_TYPE_LIST;
+    static boolean contentTypeIsNotCompatible(String acceptHeader, final MediaType actualContentType) {
+        List<MediaType> acceptableContentTypes;
+        if (StringUtils.isBlank(acceptHeader)) {
+            acceptableContentTypes = MediaTypes.GENERAL_MEDIA_TYPE_LIST;
+        } else {
+            acceptableContentTypes = new ArrayList<>();
+            String[] types = acceptHeader.split(",");
+            for (String type : types) {
+                acceptableContentTypes.addAll(getMediaTypes(type));
+            }
+        }
 
         return !Iterables.any(acceptableContentTypes, new Predicate<MediaType>() {
             @Override
@@ -330,6 +340,14 @@ public class ChannelContentResource {
                 return input.isCompatible(actualContentType);
             }
         });
+    }
+
+    private static List<MediaType> getMediaTypes(String type) {
+        try {
+            return MediaTypes.createMediaTypes(new String[]{type});
+        } catch (Exception e) {
+            return Collections.emptyList();
+        }
     }
 
 

--- a/src/test/integration/insert_and_fetch_header_spec.js
+++ b/src/test/integration/insert_and_fetch_header_spec.js
@@ -1,0 +1,35 @@
+require('./integration_config.js');
+
+var request = require('request');
+var Q = require('q');
+var channelName = utils.randomChannelName();
+var channelResource = channelUrl + "/" + channelName;
+var testName = __filename;
+
+describe(testName, function () {
+    utils.createChannel(channelName);
+
+    it('checks accept header', function (done) {
+
+        utils.postItemQ(channelResource)
+            .then(function (value) {
+                var url = value.body._links.self.href;
+                var options = {
+                    url: url,
+                    headers: {
+                        'Accept': 'text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2'
+                    }
+                };
+                request.get(options,
+                    function (err, response, body) {
+                        expect(err).toBeNull();
+                        expect(response.statusCode).toBe(200);
+                        done();
+                    });
+
+
+            })
+    });
+
+});
+

--- a/src/test/java/com/flightstats/hub/channel/ChannelContentResourceTest.java
+++ b/src/test/java/com/flightstats/hub/channel/ChannelContentResourceTest.java
@@ -1,0 +1,23 @@
+package com.flightstats.hub.channel;
+
+import org.junit.Test;
+
+import javax.ws.rs.core.MediaType;
+
+import static org.junit.Assert.assertFalse;
+
+public class ChannelContentResourceTest {
+
+    @Test
+    public void testHttpURLConnection() {
+        String accept = "text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2";
+        assertFalse(ChannelContentResource.contentTypeIsNotCompatible(accept, MediaType.WILDCARD_TYPE));
+    }
+
+    @Test
+    public void testMost() {
+        String accept = "text/html, image/gif, image/jpeg, */*; q=.2";
+        assertFalse(ChannelContentResource.contentTypeIsNotCompatible(accept, MediaType.WILDCARD_TYPE));
+    }
+
+}

